subsys/bluetooth/host/hci_core.c

```diff
@@ -2963,9 +2963,9 @@ static void hci_core_send_cmd(void)
 	err = bt_send(buf);
 	if (err) {
 		LOG_ERR("Unable to send to driver (err %d)", err);
-		k_sem_give(&bt_dev.ncmd_sem);
-		hci_cmd_done(cmd(buf)->opcode, BT_HCI_ERR_UNSPECIFIED, buf);
 		net_buf_unref(buf);
+		k_sem_give(&bt_dev.ncmd_sem);
+		hci_cmd_done(cmd(buf)->opcode, BT_HCI_ERR_UNSPECIFIED, NULL);
```

Comment:
Since you `net_buf_unref(buf);` can we not be in a case where `buf` here has 0 ref's while we do `cmd(buf)`?

Reply:
Thinking in terms of reference ownership will show that it is ok to move `buf` into `net_buf_unref` at this point.

```c
	buf = net_buf_get(&bt_dev.cmd_tx_queue, K_NO_WAIT); // ref moved into `buf`

	bt_dev.sent_cmd = net_buf_ref(buf); // borrow. A clone is moved into `bt_dev.sent_cmd`.

	err = bt_send(buf); // borrow if err
	if (err) {
		net_buf_unref(buf); // moved here
		// No use of `buf` beyond this point
		k_sem_give(&bt_dev.ncmd_sem);
		hci_cmd_done(cmd(buf)->opcode, BT_HCI_ERR_UNSPECIFIED, NULL);
		bt_tx_irq_raise();
	}
```

If you are wondering where `hci_cmd_done` will get a reference from, it's from `bt_dev.sent_cmd` which carries a reference.

Reply:
But doesn't this still increase the risk of a race condition?

If you remove the reference here, then it ref count may be 0 before `hci_cmd_done(cmd(buf)->opcode`, and in that time frame couldn't another thread take the buffer for something else since the ref count is 0? 

You write `// No use of `buf` beyond this point` but do `hci_cmd_done(cmd(buf)->opcode` 2 lines below.

Reply:
> You write `// No use of `buf` beyond this point` but do `hci_cmd_done(cmd(buf)->opcode` 2 lines below.

You are right. This is an error.